src/Lean/Elab/Tactic/DiscrTreeKey.lean

```diff
@@ -0,0 +1,82 @@
+/-
+Copyright (c) 2024 Matthew Robert Ballard. All rights reserved.
+Released under Apache 2.0 license as described in the file LICENSE.
+Authors: Tomas Skrivan, Matthew Robert Ballard
+-/
+prelude
+import Init.Tactics
+import Lean.Elab.Command
+import Lean.Meta.Tactic.Simp.SimpTheorems
+
+namespace Lean.Elab.Tactic.DiscrTreeKey
+
+open Lean.Meta DiscrTree
+open Lean.Elab.Tactic
+open Lean.Elab.Command
+
+private def mkKey (e : Expr) (simp : Bool) : MetaM (Array Key) := do
+  let (_, _, type) ← withReducible <| forallMetaTelescopeReducing e
+  let type ← whnfR type
+  if simp then
+    match type.eq? with
+    | some (_, lhs, _) => mkPath lhs simpDtConfig
+    | none => throwError "unexpected kind of 'simp' theorem{indentExpr type}"
```

Comment:
```suggestion
    match type.eq? with
    | some (_, lhs, _) => mkPath lhs simpDtConfig
    | none => throwError "unexpected kind of 'simp' theorem{indentExpr type}"
```

Simp theorems can have other forms, I think. They support `Iff` lemmas. And other lemmas can be `simp` as well, they are implicit converted to `… = True`.  Not sure if you want to support that. See the private `preprocess` used by `mkSimpTheoremsFromConst`.

If you plan to not support that, maybe leave a comment that this is intentional.

Reply:
Good point. I'll look into this